eventqueue: add name field, additional unit test coverage

ianvernon opened the pull request:
Provide a means of differentiating logs for different `EventQueues` by allowing creators of  `EventQueue` to provide a name. Update the printing of statistics for a given event being processed to reference the name of the queue that is processing / has processed the event. This field will make it easier to track event processing in relation to a given owner of a queue (policy, endpoint, etc.). 

Also add additional unit tests; the package now has 100 % code coverage.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7469)
<!-- Reviewable:end -->

Commit:
eventqueue: add name field, additional logging for debugging

Provide a means of differentiating logs for different `EventQueues` by allowing creators of
`EventQueue` to provide a name. Update the printing of statistics for a given event being
processed to reference the name of the queue that is processing / has processed the event.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
eventqueue: add misc. unit tests

Coverage is now 100% for this package.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/eventqueue/eventqueue.go:
```diff
@@ -121,6 +138,12 @@ type eventStatistics struct {
 }
 
+// UpdateName updates the name of the EventQueue.
+func (q *EventQueue) UpdateName(name string) {
```
A name of `EventQueue` should never change during its execution, a name should be part of the struct's identity. If you really really want the ability to change the name then you need to protect it against concurrent access.

ianvernon replied:
I'll add a mutex which protects this field. The reason this has to be done is because I want to name the queue based off of the endpoint ID, but the endpoint ID is not allocated as part of endpoint creation - it is done by the endpointmanager, which is done after the endpoint's EventQueue is initialized.

ianvernon replied:
Added a mutex to protect the field in a follow-up commit.

joestringer replied:
How much work is it to shift the endpointID allocation to before the endpoint initialization? It seems strange to me that something so fundamental to the endpoint is actually considered mutable state that's being modified after the endpoint object is initialized.

ianvernon replied:
🤔 I checked out the code and I see no reason why we can't move the endpoint ID allocation to endpoint creation. I'll play around with that.

tgraf replied:
Be advised that the main path is:
```
func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.EndpointChangeRequest) (*endpoint.Endpoint, error) {

        ep, err := endpoint.NewEndpointFromChangeModel(epTemplate)
```

The endpoint ID will *not* be set in the template. It is assigned on insertion into the manager. The plugin calling the API is no longer responsible to allocate the endpoint ID. Allocating earlier is possible but it will complicate the endpoint ID release mechanism in all of the error paths.

ianvernon replied:
@tgraf agreed, I came to the conclusion that allocating upon endpoint creation will complicate things as well. We'd have to ensure that all error paths between endpoint initialization and insertion into the endpointmanager correctly release the ID for the endpoint, which is not good. The correct solution is to start the queue as the endpoint is inserted into the endpointmanager (i.e., right after it has gotten an ID).

joestringer replied:
That sounds like a fair compromise.

Commit:
eventqueue: add logic to gracefully handle nil queues

Given that the endpoint's EventQueue is available to endpoint consumers, but is
not instantiated until it is allocated an ID in the endpointmanager, add checks
for if the queue is nil for \`Run\`, \`Enqueue\`, and \`Stop\`. This way, if an
endpoint is deleted before its EventQueue is created / started by the
endpointmanager, it does not have to worry about whether the queue is started or
stopped yet.

Signed-off by: Ian Vernon <ian@cilium.io>
